### PR TITLE
tests/monitoring/monitoring: Replace update with Patch

### PR DESCRIPTION
**Pull request by avlitman**
Switched from Update to Patch to avoid overwriting resource fields
unintentionally. Patch allows precise updates to specific fields,
reducing the risk of accidental changes and improving test reliability.

### What this PR does
Before this PR:
The test using update

After this PR:
The test using patch

jira-ticket: https://issues.redhat.com/browse/CNV-44576

### Release note

```release-note 
none

```

### CODE REVIEW

File: tests/monitoring/monitoring.go [hunk 1]
```diff
@@ -124,19 +126,24 @@ var _ = Describe("[Serial][sig-monitoring]Monitoring", Serial, decorators.SigMon
...
+							},
+						},
 					},
 				},
 			}
 
+			patchBytes, err := json.Marshal(patchData)
+			Expect(err).ToNot(HaveOccurred())
+
 			Eventually(func() error {
```
Comment (nunnatsa) [inline tests/monitoring/monitoring.go:145]: I would use json patch in this case. e.g.
```suggestion
			patchBytes, err := patch.New(WithAdd("/spec/template/nodeSelector/kubernetes.io~1hostname", "does-not-exist").GeneratePayload()
			Expect(err).ToNot(HaveOccurred())

			_, err = virtClient.KubeVirt(originalKv.Namespace).Patch(context.Background(), originalKv.Name, types.JSONPatchType, patchBytes, metav1.PatchOptions{})			
```

(note: `~1` is the go path replacement for "/")

Comment (nunnatsa) [reply tests/monitoring/monitoring.go:145]: The whole idea is to get rid of the Eventually.

File: tests/monitoring/monitoring.go [hunk 2]
```diff
@@ -160,10 +167,13 @@ var _ = Describe("[Serial][sig-monitoring]Monitoring", Serial, decorators.SigMon
 			kv, err := virtClient.KubeVirt(originalKv.Namespace).Get(context.Background(), originalKv.Name, metav1.GetOptions{})
 			Expect(err).ToNot(HaveOccurred())
 
-			kv.Spec.CustomizeComponents = v1.CustomizeComponents{}
+			patchData := map[string]interface{}{}
+
+			patchBytes, err := json.Marshal(patchData)
+			Expect(err).ToNot(HaveOccurred())
```
Comment (nunnatsa) [inline tests/monitoring/monitoring.go:173]: Please use json patch; e.g.
```suggestion
            patchBytes, err := patch.New(WithRemove("/spec/customizeComponents").GeneratePayload()
			Expect(err).ToNot(HaveOccurred())
			patchData := map[string]interface{}{}
```

File: tests/monitoring/monitoring.go [hunk 3]
```diff
@@ -160,10 +167,13 @@ var _ = Describe("[Serial][sig-monitoring]Monitoring", Serial, decorators.SigMon
...
 
 			Eventually(func() error {
-				kv, err = virtClient.KubeVirt(originalKv.Namespace).Update(context.Background(), kv, metav1.UpdateOptions{})
+				_, err = virtClient.KubeVirt(originalKv.Namespace).Patch(context.Background(), originalKv.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
```
Comment (nunnatsa) [inline tests/monitoring/monitoring.go:176]: Please use json type.

Please remove the `Eventually`

**Review by nunnatsa — COMMENTED** [review]
Please use json patch to be more accurate, and use less code. Also, please get rid of the `Eventually` - we don't need them if we're using patch instead of update. details are inline.

File: tests/monitoring/monitoring.go [hunk 4]
```diff
@@ -160,12 +154,17 @@ var _ = Describe("[Serial][sig-monitoring]Monitoring", Serial, decorators.SigMon
 			kv, err := virtClient.KubeVirt(originalKv.Namespace).Get(context.Background(), originalKv.Name, metav1.GetOptions{})
 			Expect(err).ToNot(HaveOccurred())
 
-			kv.Spec.CustomizeComponents = v1.CustomizeComponents{}
+			patchData := []map[string]interface{}{
```
Comment (codingben) [inline tests/monitoring/monitoring.go:157]: Can you please use functions from `patch.go`? For example:
```
patchData, err := patch.New(patch.WithRemove("/spec/customizeComponents")).GeneratePayload()
Expect(err).ToNot(HaveOccurred())
```

File: tests/monitoring/monitoring.go [hunk 5]
```diff
@@ -124,33 +126,25 @@ var _ = Describe("[Serial][sig-monitoring]Monitoring", Serial, decorators.SigMon
...
-			kv.Spec.CustomizeComponents = v1.CustomizeComponents{
-				Patches: []v1.CustomizeComponentsPatch{
-					{
-						ResourceName: virtHandler.deploymentName,
-						ResourceType: "DaemonSet",
-						Patch:        `{"spec":{"template":{"spec":{"nodeSelector":{"kubernetes.io/hostname":"does-not-exist"}}}}}`,
-						Type:         v1.StrategicMergePatchType,
-					},
+
+			patchData := []map[string]interface{}{
```
Comment (codingben) [inline tests/monitoring/monitoring.go:130]: Also here, I'd suggest to use functions from `patch.go`:
```
patchData, err := patch.New(patch.WithAdd("/spec/template/spec/nodeSelector/kubernetes.io~1hostname", "does-not-exist")).GeneratePayload()
Expect(err).ToNot(HaveOccurred())
```

File: tests/monitoring/monitoring.go [hunk 6]
```diff
@@ -21,6 +21,7 @@ package monitoring
```
Comment (dankenigsberg) [inline tests/monitoring/monitoring.go:21]:
```
tests/monitoring/monitoring: Replace update with Patch
Signed-off-by: avlitman <alitman@redhat.com>
```
In your commit message, please explain *why* this commit is important. What is wrong with repeatedly Updating?

Comment (avlitman) [reply tests/monitoring/monitoring.go:21]: Done

File: tests/monitoring/monitoring.go (hunk 4 above)
Comment (avlitman) [reply tests/monitoring/monitoring.go]: Done

File: tests/monitoring/monitoring.go (hunk 5 above)
Comment (avlitman) [reply tests/monitoring/monitoring.go]: Done

File: tests/monitoring/monitoring.go [hunk 7]
```diff
@@ -124,33 +126,18 @@ var _ = Describe("[Serial][sig-monitoring]Monitoring", Serial, decorators.SigMon
...
+			Expect(err).ToNot(HaveOccurred())
 
-			Eventually(func() string {
-				vh, err := virtClient.AppsV1().DaemonSets(originalKv.Namespace).Get(context.Background(), virtHandler.deploymentName, metav1.GetOptions{})
-				Expect(err).ToNot(HaveOccurred())
-				return vh.Spec.Template.Spec.NodeSelector[k8sv1.LabelHostname]
-			}, 90*time.Second, 5*time.Second).Should(Equal("does-not-exist"))
+			vh, err := virtClient.AppsV1().DaemonSets(originalKv.Namespace).Get(context.Background(), virtHandler.deploymentName, metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(vh.Spec.Template.Spec.NodeSelector[k8sv1.LabelHostname]).To(Equal("does-not-exist"))
```
Comment (nunnatsa) [inline tests/monitoring/monitoring.go:138]: please keep this eventually. it may take some time to update

File: tests/monitoring/monitoring.go [hunk 8]
```diff
@@ -124,33 +126,18 @@ var _ = Describe("[Serial][sig-monitoring]Monitoring", Serial, decorators.SigMon
...
 
-			Eventually(func() int {
-				vh, err := virtClient.AppsV1().DaemonSets(originalKv.Namespace).Get(context.Background(), virtHandler.deploymentName, metav1.GetOptions{})
-				Expect(err).ToNot(HaveOccurred())
-				return int(vh.Status.NumberAvailable + vh.Status.NumberUnavailable)
-			}, 90*time.Second, 5*time.Second).Should(Equal(0))
+			Expect(int(vh.Status.NumberAvailable + vh.Status.NumberUnavailable)).To(Equal(0))
```
Comment (nunnatsa) [inline tests/monitoring/monitoring.go:140]: same

Comment (avlitman) [reply tests/monitoring/monitoring.go]: reverted back to use eventually

File: tests/monitoring/monitoring.go (hunk 7 above)
Comment (avlitman) [reply tests/monitoring/monitoring.go]: reverted back to use eventually

File: tests/monitoring/monitoring.go [hunk 9]
```diff
@@ -124,21 +126,9 @@ var _ = Describe("[Serial][sig-monitoring]Monitoring", Serial, decorators.SigMon
...
-					},
-				},
-			}
 
-			Eventually(func() error {
-				kv, err = virtClient.KubeVirt(originalKv.Namespace).Update(context.Background(), kv, metav1.UpdateOptions{})
-				return err
-			}, 30*time.Second, 1*time.Second).ShouldNot(HaveOccurred())
+			_, err = patch.New(patch.WithAdd("/spec/template/spec/nodeSelector/kubernetes.io~1hostname", "does-not-exist")).GeneratePayload()
+			Expect(err).ToNot(HaveOccurred())
```
Comment (nunnatsa) [inline tests/monitoring/monitoring.go:131]: I think you accidentally deleted the actual Patch

File: tests/monitoring/monitoring.go [hunk 10]
```diff
@@ -124,21 +126,9 @@ var _ = Describe("[Serial][sig-monitoring]Monitoring", Serial, decorators.SigMon
...
-						Type:         v1.StrategicMergePatchType,
-					},
-				},
-			}
 
-			Eventually(func() error {
-				kv, err = virtClient.KubeVirt(originalKv.Namespace).Update(context.Background(), kv, metav1.UpdateOptions{})
-				return err
-			}, 30*time.Second, 1*time.Second).ShouldNot(HaveOccurred())
+			_, err = patch.New(patch.WithAdd("/spec/template/spec/nodeSelector/kubernetes.io~1hostname", "does-not-exist")).GeneratePayload()
```
Comment (codingben) [inline tests/monitoring/monitoring.go:130]: It should return patch bytes (`patchBytes, err = ...`) so then you can send in `PATCH` (HTTP request) with this actual data. Very similar to what you did below:
```
patchBytes, err := patch.New(patch.WithRemove("/spec/customizeComponents")).GeneratePayload()
Expect(err).ToNot(HaveOccurred())

_, err = virtClient.KubeVirt(originalKv.Namespace).Patch(context.Background(), originalKv.Name, types.JSONPatchType, patchBytes, metav1.PatchOptions{})
Expect(err).ToNot(HaveOccurred())
```
Here `patchBytes` is used when you do `Patch` API request, instead `Update`.

Comment (fossedihelm) [inline tests/monitoring/monitoring.go:130]: I think this is wrong! Other than that no patch is being performed, with this you are going to patch `/spec/template/spec/nodeSelector/kubernetes.io~1hostname` which actually does not exist.
The patch that should be performed is the addition of the `kv.spec.customizedComponents`:
It should be something like this:
```
originalKv := util.GetCurrentKv(virtClient)
customizedComponents := v1.CustomizeComponents{
	Patches: []v1.CustomizeComponentsPatch{
		{
			ResourceName: virtHandler.deploymentName,
			ResourceType: "DaemonSet",
			Patch:        `{"spec":{"template":{"spec":{"nodeSelector":{"kubernetes.io/hostname":"does-not-exist"}}}}}`,
			Type:         v1.StrategicMergePatchType,
		},
	},
}
patchBytes, err := patch.New(patch.WithAdd("/spec/customizeComponents", customizedComponents)).GeneratePayload()
Expect(err).ToNot(HaveOccurred())
kv, err := virtClient.KubeVirt(originalKv.Namespace).Patch(context.Background(), originalKv.Name, types.JSONPatchType, patchBytes, metav1.PatchOptions{})
Expect(err).ToNot(HaveOccurred())
```
In this way, you are patching the KubeVirt resource, which will patch the virt-handler with the patch inside the customizedComponent.

File: tests/monitoring/monitoring.go [hunk 11]
```diff
@@ -160,12 +150,11 @@ var _ = Describe("[Serial][sig-monitoring]Monitoring", Serial, decorators.SigMon
 			kv, err := virtClient.KubeVirt(originalKv.Namespace).Get(context.Background(), originalKv.Name, metav1.GetOptions{})
```
Comment (fossedihelm) [inline tests/monitoring/monitoring.go:150]: I think we can get rid of this `Get`

**Review by fossedihelm — CHANGES_REQUESTED** [review]
Thank you @avlitman!

File: tests/monitoring/monitoring.go [hunk 12]
```diff
@@ -134,49 +134,43 @@ var _ = Describe("[Serial][sig-monitoring]Monitoring", Serial, decorators.SigMon
...
 				Expect(err).ToNot(HaveOccurred())
 				return vh.Spec.Template.Spec.NodeSelector[k8sv1.LabelHostname]
-			}, 90*time.Second, 5*time.Second).Should(Equal("does-not-exist"))
+			}).WithTimeout(90 * time.Second).WithPolling(5 * time.Second).Should(Equal("does-not-exist"))
 
 			Eventually(func() int {
 				vh, err := virtClient.AppsV1().DaemonSets(originalKv.Namespace).Get(context.Background(), virtHandler.deploymentName, metav1.GetOptions{})
 				Expect(err).ToNot(HaveOccurred())
 				return int(vh.Status.NumberAvailable + vh.Status.NumberUnavailable)
-			}, 90*time.Second, 5*time.Second).Should(Equal(0))
```
Comment (machadovilaca) [inline tests/monitoring/monitoring.go:153]: in the codebase most often than not, the previous approach is used, why the change?

File: tests/monitoring/monitoring.go [hunk 13]
```diff
@@ -120,11 +122,9 @@ var _ = Describe("[Serial][sig-monitoring]Monitoring", Serial, decorators.SigMon
 	})
 
 	Context("System Alerts", func() {
-		disableVirtHandler := func() *v1.KubeVirt {
+		disableVirtHandler := func() {
 			originalKv := util.GetCurrentKv(virtClient)
```
Comment (machadovilaca) [inline tests/monitoring/monitoring.go:126]: since now we only use the name and the namespace, can you check if we can avoid this call?

File: tests/monitoring/monitoring.go [hunk 14]
```diff
@@ -134,49 +134,43 @@ var _ = Describe("[Serial][sig-monitoring]Monitoring", Serial, decorators.SigMon
...
 		}
 
-		restoreVirtHandler := func(kv *v1.KubeVirt) {
+		restoreVirtHandler := func() {
 			originalKv := util.GetCurrentKv(virtClient)
-			kv, err := virtClient.KubeVirt(originalKv.Namespace).Get(context.Background(), originalKv.Name, metav1.GetOptions{})
-			Expect(err).ToNot(HaveOccurred())
 
-			kv.Spec.CustomizeComponents = v1.CustomizeComponents{}
+			patchBytes, err := patch.New(patch.WithRemove("/spec/customizeComponents")).GeneratePayload()
```
Comment (machadovilaca) [inline tests/monitoring/monitoring.go:158]: nice :rocket:

**Review by machadovilaca — COMMENTED** [review]
good job, just a couple questions

File: tests/monitoring/monitoring.go (hunk 12 above)
Comment (nunnatsa) [reply tests/monitoring/monitoring.go:153]: This is something I'm trying to change. slowly...

Comment (nunnatsa) [reply tests/monitoring/monitoring.go:153]: The Eventually/Constantly old API is not the best, and not consistent. I think we must move to the more explicit version of it.

Personally, I'm changing it whenever I can.

File: tests/monitoring/monitoring.go (hunk 14 above)
Comment (avlitman) [reply tests/monitoring/monitoring.go:158]: so util.GetCurrentKv is calling util.GetKvList(virtClient). I think it's better to leave as is

File: tests/monitoring/monitoring.go (hunk 13 above)
Comment (avlitman) [reply tests/monitoring/monitoring.go]: so util.GetCurrentKv is calling util.GetKvList(virtClient). I think it's better to leave as is
